drift/lib/src/runtime/query_builder/expressions/case_when.dart

```diff
@@ -2,35 +2,113 @@ import 'package:meta/meta.dart';
 
 import '../query_builder.dart';
 
-/// A `CASE WHEN` expression in sqlite.
+/// `CASE WHEN` expression **without** base expression in sqlite.
 ///
-/// This class supports when expressions with or without a base expression.
+/// This expression evaluates to the first value in [BaseCaseWhenExpression.orderedCases]
+/// for which its key evaluates to `true`.
+///
+/// See [BaseCaseWhenExpression] for more info.
+class CaseWhenExpression<R extends Object>
+    extends BaseCaseWhenExpression<bool, R> {
+  /// Creates a `CASE WHEN` expression from the independent components.
+  ///
+  /// This expression evaluates to the first value in [cases]
+  /// for which its key evaluates to `true`.
+  ///
+  /// If [cases] is empty - throws [ArgumentError]
+  CaseWhenExpression({
+    required List<CaseWhen<bool, R>> cases,
+    Expression<R>? orElse,
+  }) : super._(null, cases, orElse);
+}
+
+/// `CASE WHEN` expression **with** base expression in sqlite.
+///
+/// For internal use. Must be used via [Expression.caseMatch]
+///
+/// See [BaseCaseWhenExpression] for more info.
+@internal
+class CaseWhenExpressionWithBase<T extends Object, R extends Object>
+    extends BaseCaseWhenExpression<T, R> {
+  /// Creates a `CASE WHEN` expression from the independent components.
+  ///
+  /// If [cases] is empty - throws [ArgumentError]
+  CaseWhenExpressionWithBase(
+    Expression<T> base, {
+    required List<CaseWhen<T, R>> cases,
+    Expression<R>? orElse,
+  }) : super._(base, cases, orElse);
+}
+
+/// A single when-then case for `CASE WHEN` expression in sqlite.
+///
+/// TODO check if it should be replaced with multiple return values.
+///
+/// [Multiple return values GitHub issue](https://github.com/dart-lang/language/issues/68)
+class CaseWhen<T extends Object, R extends Object> {
+  /// Expression to use after `WHEN`
+  final Expression<T> when;
+
+  /// Expression to use after `THEN`
+  final Expression<R> then;
+
+  /// Creates a single case with expression for when and expression for then.
+  const CaseWhen(this.when, {required this.then});
```

Comment:
then is marked as a named parameter to mark explicitly that the last expression specified in constructor is used after `THEN`